Feature: Format counties

emilyliu7321 opened the pull request:
## Description

Builds on @anthonyjpesce's existing PR. This PR adds formatting for county-level results and some other clean-up:

- County-level formatting (generalizes + refactors some code to handle both precinct and county subunits)
- Handle Clarity data when only one `Contest` is present 
- Adds fixture for state-level GA results for only two contests at `tests/fixtures/2020-11-03_GA.xml` 
- Adds fixture for precinct-level GA results with only one contest at `tests/fixtures/atkinson_precincts_2020-11-03_GA_G_P.xml` 

Currently, county IDs are formatted as `{fips}_{county_name}`, e.g. `13217_newton` to match the way we're formatting precinct names, but we can change this to just be the fips code to match our ap/edison formatters too. 

## Test Steps

``elexclarity 105369 GA --level=county``
``elexclarity 105369 GA --level=county --filename="tests/fixtures/2020-11-03_GA.xml"``

Produces something formatted like so: 

```
"State Senate District 1": {
    "source": "clarity",
    "name": "State Senate District 1",
    "precinctsReportingPct": 100.0,
    "subunits": {
      "13029_bryan": {
        "id": "13029_bryan",
        "counts": {
          "ben-watson-i-rep": 17287
        }
      },
      "13051_chatham": {
        "id": "13051_chatham",
        "counts": {
          "ben-watson-i-rep": 44070
        }
      },
      "13179_liberty": {
        "id": "13179_liberty",
        "counts": {
          "ben-watson-i-rep": 10835
        }
      }
    },
    "counts": {
      "ben-watson-i-rep": 72192
    }
  }
```
 
README is updated with some sample commands.

anthonyjpesce commented on README.rst:
```diff
@@ -7,6 +7,23 @@ A CLI tool for pulling in election results from sites using Clarity.
+Pulling data from Clarity:
+
+* ``elexclarity 105369 GA --outputType=summary``
+* ``eelexclarity 105369 GA --outputType=settings``
```
small typo here

anthonyjpesce reviewed (commented):
This looks great, thank you for adding the county option! I feel a little bad having those first few things in the readme if we haven't implemented them yet -- I guess we should probably make a ticket for those. Also, I think we need some unit tests for the county stuff.

emilyliu7321 commented:
@anthonyjpesce Thanks for the review! The output for `summary` and `settings` is part of Susan's branch, which is the base for both of ours, so I pushed a small commit that gets that working here too and added clarification in the README that those two `outputType`s still need formatting. 

Added county unit tests too!

anthonyjpesce reviewed (approved):
This LGTM, thank you! We should flag @st421 before merge though

st421 commented on src/elexclarity/client.py:
```diff
@@ -125,7 +126,7 @@ def get_results(self, electionid, statepostal, level, **kwargs):
                 results.append(self.get_county_results(statepostal, county, **kwargs))
             return results
-        elif level == "state":
+        elif level == "state" or 'county':
```
This needs to be two separate equality checks (or a `level in [...]`). Source:

```
>>> level="precinct"
>>> print(level == "state" or 'county')
county
```

st421 commented on src/elexclarity/formatters.py:
```diff
@@ -18,16 +18,17 @@ def get_id_for_choice(self, name):
-        name = slugify(name)
+        name = slugify(subunit_name)
+        if fips is None:
+            fips = self.county_lookup.get(subunit_name)
```
I am probably misunderstanding, but tracing this back it seems like we've already done the FIPS lookup but are doing it again here when level == county?

emilyliu7321 replied:
Yep, the way Anthony's PR had this set up was when you're handling county-level XML data to parse precincts, you pass down a county fips code to all of its affiliated precincts because you can't access the county name in the <Precinct/> tag, but when handling state-level XML, you access the county names directly in the <County/> tag

st421 reviewed (approved):
You fixed the one actual bug; the other comment can be addressed later
